Content.Server/Paper/PaperSystem.cs

```diff
@@ -37,15 +39,35 @@ private void OnInit(EntityUid uid, PaperComponent paperComp, ComponentInit args)
             paperComp.Mode = PaperAction.Read;
             UpdateUserInterface(uid, paperComp);
 
-            if (TryComp<AppearanceComponent>(uid, out var appearance))
+            UpdateAppearance(uid, paperComp);
+
+        }
+
+        private void OnMapInit(EntityUid uid, PaperComponent paperComp, MapInitEvent args)
+        {
+            if (paperComp.UseLocale && Loc.TryGetString(paperComp.Content, out var locString))
             {
-                if (paperComp.Content != "")
-                    appearance.SetData(PaperVisuals.Status, PaperStatus.Written);
+                paperComp.Content = locString;
+            }
 
-                if (paperComp.StampState != null)
-                    appearance.SetData(PaperVisuals.Stamp, paperComp.StampState);
+            if (paperComp.Content.Length > paperComp.ContentSize)
+            {
+                paperComp.Content = paperComp.Content[..paperComp.ContentSize];
             }
+        }
+
+        private void UpdateAppearance(EntityUid uid, PaperComponent paperComp)
```

Comment:
this should have a nullable AppearanceComponent in the parameters. You could then just resolve it and remove the hascomp and not resolve it in SetData multiple times (like i mentioned earlier)